hushline/crypto.py

```diff
@@ -1,21 +1,72 @@
 import os
+from base64 import urlsafe_b64decode, urlsafe_b64encode
 
 from cryptography.fernet import Fernet
+from cryptography.hazmat.primitives.kdf.scrypt import Scrypt
 from flask import current_app
 from pysequoia import Cert, encrypt
 
+
+def generate_salt() -> str:
+    """
+    Generate a random salt for use in encryption key derivation.
+    """
+    return urlsafe_b64encode(os.urandom(16)).decode()
+
+
+def get_encryption_key(scope: bytes | str | None = None, salt: str | None = None) -> Fernet:
+    """
+    Return the default Fernet encryption key. If a scope and salt are provided, a unique encryption
+    key will be derived based on the scope and salt.
+    """
+    encryption_key = os.environ.get("ENCRYPTION_KEY")
+    if encryption_key is None:
+        raise ValueError("Encryption key not found. Please check your .env file.")
+
+    # If a scope is provided, we will use it to derive a unique encryption key
+    if scope is not None and salt is not None:
+        # Convert the scope to bytes if it is a string
+        if isinstance(scope, str):
+            scope_bytes = scope.encode()
+        elif isinstance(scope, bytes):
+            scope_bytes = scope
+
+        # Convert the encryption key and salt to bytes
+        encryption_key_bytes = urlsafe_b64decode(encryption_key)
+        salt_bytes = urlsafe_b64decode(salt)
+
+        # Use Scrypt to derive a unique encryption key based on the scope
+        kdf = Scrypt(
+            salt=salt_bytes,
+            length=32,
+            n=2**14,
+            r=8,
+            p=1,
```

Comment:
should we extract these literals to constants?

Reply:
Implemented in https://github.com/scidsg/hushline/pull/515/commits/740aa15eded55e3d699124a6a2752b77cf4eb92b